### DISCUSSION TITLE
Annotation Ink & Text Not Captured When Printing

### DISCUSSION THREAD

**Discussion by cristianscode** [stephanrauh/ngx-extended-pdf-viewer#1635] [Q&A]
Hi, amazing project. I was just wondering if there is a way to capture the annotation (ink & text) layer when printing? This would be useful since the ink layer can be used for signatures! Which would be such an amazing feature!!!!

**Comment by stephanrauh** [comment]
I suppose you're talking about the brand-new editor? I'm not sure, but your question might be related to https://github.com/stephanrauh/ngx-extended-pdf-viewer/issues/1592. My working theory is I ran into a merge conflict and solved it the wrong way. As things go, I've started a few minutes ago investigating the topic. It's a nasty surprise to learn that printing is affected, too.

**Reply by cristianscode** [reply]
Yikes! So this feature did work in the past? Let me try installing a previous version and see if that is the case!

**Reply by stephanrauh** [reply]
No idea. :) The first time I tried, it was broken. Adding insult to injury, I don't have a clue where the Mozilla guys implemented the feature. But wait - it's good we're chatting about it. You've inspired me to have a look at the Git history, and I've found this: https://github.com/mozilla/pdf.js/pull/15047. Maybe it helps us to figure out what I'm doing wrong!

**Reply by cristianscode** [reply]
Great! I've downgraded to version 15 and I don't see an annotation feature hence the reason you call it "the brand-new editor". If the is anything I can do to help you out let me know. I can test as well, just let me know :)

**Reply by stephanrauh** [reply]
Brand-new means that the Mozilla team has added the first commit on June, 4th. I always tried to support the feature, but I hid it until I considered it mature. Set `[showEditor]="true"` (and remind me to document the attribute). :)

**Reply by cristianscode** [reply]
I've set [showEditor]="true" and no luck. I am currently using the 16.0.0-alpha.0 version

**Reply by stephanrauh** [reply]
As for helping: currently I'm debugging the original pdf.js project and trying to compare it to my own fork. I'm both tired and a bit ill, so I guess you can really help me, if you want to clone pdf.js and get it up and running. It's a matter of 10, 15 minutes, I guess - the walk-through is pretty good: https://github.com/mozilla/pdf.js#getting-the-code

**Reply by stephanrauh** [reply]
Yeah, I've added the feature toggle three months ago. Maybe you've got an older version.

**Reply by stephanrauh** [reply]
Hm. There's no point in cloning pdf.js locally. It's easier to debug it here: https://mozilla.github.io/pdf.js/web/viewer.html

**Reply by cristianscode** [reply]
I think I'm up to date, I'm trying to set up pdf.js I just installed it locally

**Reply by cristianscode** [reply]
Shucks! I'll use the only viewer then

**Reply by stephanrauh** [reply]
It seems you'll also need Chrome. All the magic happens in the worker thread, and I didn't manage to convince Firefox to add breakpoints over there. But of course, it's possible the root problem is lurking in the main thread.

**Reply by cristianscode** [reply]
I'm trying to debug using Chrome's console. Are you still working on it as well?

**Reply by stephanrauh** [reply]
I am. But more likely than not, I'm calling it a day soon. Although it's currently exciting. :) I've found something: pdf.js calls `AnnotationFactory.printNewAnnotations()`, but my fork seems to skip it. Remains the question why that's so.

**Reply by cristianscode** [reply]
Ahh! Sounds like that function could be important haha. I understand if you have to call it a day! I appreciate the fast responses, hopefully we can solve this. I will continue to look at it, but if you solve it please let me know as well. Happy Holidays!!

**Reply by stephanrauh** [reply]
I'm giving up now. My latest insight looks promising. The error is visible at the main thread, in the file `app.js`. The method `beforePrint()` shows a difference. `this.pdfDocument?.annotationStorage.print` has a private attribute called `#serializable`. It's an empty map in ngx-extended-pdf-viewer, but it contains one entry in pdf.js. It's a Bezier path - which matches my test case, so I hope that's the squiggly line I've added to the PDF file.

**Reply by stephanrauh** [reply]
Original pdf.js:
![grafik](https://user-images.githubusercontent.com/3045767/209005911-42d35b73-3c77-477a-b1d1-f4851c2df0c5.png)

My fork:
![grafik](https://user-images.githubusercontent.com/3045767/209006245-4607cb35-bb2a-4c98-930b-ec0c99f78321.png)

**Reply by cristianscode** [reply]
Ohhhhhh I see!!

**Reply by cristianscode** [reply]
So would AnnotationFactory.printNewAnnotations() assign #serializable?

**Reply by stephanrauh** [reply]
No, I don't think so. I've worked my way up the call stack. `printNewAnnotations()` was the first difference I've spotted, deep down in the depths of the worker thread. At the moment, I believe the difference is a lot closer to the UI layer.

**Reply by cristianscode** [reply]
Awh man, seems like its going to be a complicated fix. I was trying to look into it but no luck on my side :(

**Reply by stephanrauh** [reply]
Found it! I've modified an internal API, and that's something no merge tool can detect.

**Reply by stephanrauh** [reply]
I can't promise to publish the bugfix before Christmas, but we're close. Thanks for your help, @cristianscode. Maybe you think you didn't do much, but motivate me you did, and that's no small feat!

**Reply by cristianscode** [reply]
That's amazing! I've never collaborated with someone on Github. Talking to you and debugging this thing makes me want to get more involved in this community! No rush on pushing the bugfix just let me know when you deploy so I can test with you. Hope to hear from you soon, Happy Holidays!!

**Reply by stephanrauh** [reply]
Your bug fix has landed with version 16.0.0-alpha.1. Unfortunately, now my copy of Chrome doesn't print at all now, but I don't think that's related to the bug fix. It works fine when downloading the file or using Firefox.

**Comment by cristianscode** [comment]
I finally was able to test it!!!! ANDDDD it worked, thank you so much!! Love this project
